pageserver/src/tenant/storage_layer/remote_layer.rs

```diff
@@ -129,12 +129,12 @@ impl PersistentLayer for RemoteLayer {
         None
     }
 
-    fn iter(&self, _ctx: &RequestContext) -> Result<LayerIter<'_>> {
-        bail!("cannot iterate a remote layer");
+    fn load_val_refs(&self, _ctx: &RequestContext) -> Result<Vec<(Key, Lsn, super::ValueRef)>> {
+        bail!("cannot load value refs of a remote layer");
     }
 
-    fn key_iter(&self, _ctx: &RequestContext) -> Result<LayerKeyIter<'_>> {
-        bail!("cannot iterate a remote layer");
+    fn load_keys(&self, _ctx: &RequestContext) -> Result<Vec<(Key, Lsn, u64)>> {
+        bail!("cannot load keys of a remote layer");
```

Comment:
In future it might make sense to drop these altogether from the trait, since they are not really trait methods if only one implementation supports them. Could just downcast to expected type instead.

Reply:
I think such upcasting needs the [`dyn` upcasting feature](https://github.com/rust-lang/rfcs/blob/master/text/3324-dyn-upcasting.md), no?